tools/build_pytorch_libs.py

```diff
@@ -0,0 +1,283 @@
+from .setup_helpers.env import (IS_ARM, IS_DARWIN, IS_LINUX, IS_PPC, IS_WINDOWS,
+                                DEBUG, REL_WITH_DEB_INFO, USE_MKLDNN,
+                                check_env_flag, check_negative_env_flag, hotpatch_build_env_vars)
+
+import os
+import sys
+import distutils
+import distutils.sysconfig
+from distutils.file_util import copy_file
+from distutils.dir_util import copy_tree
+from subprocess import check_call, call, check_output
+from distutils.version import LooseVersion
+from .setup_helpers.cuda import USE_CUDA, CUDA_HOME
+from .setup_helpers.dist_check import USE_DISTRIBUTED, USE_GLOO_IBVERBS
+from .setup_helpers.nccl import USE_SYSTEM_NCCL, NCCL_INCLUDE_DIR, NCCL_ROOT_DIR, NCCL_SYSTEM_LIB
+from .setup_helpers.rocm import ROCM_HOME, ROCM_VERSION, USE_ROCM
+from .setup_helpers.nnpack import USE_NNPACK
+from .setup_helpers.qnnpack import USE_QNNPACK
+from .setup_helpers.cudnn import CUDNN_INCLUDE_DIR, CUDNN_LIB_DIR, CUDNN_LIBRARY, USE_CUDNN
+
+
+from pprint import pprint
+from glob import glob
+import multiprocessing
+import shutil
+
+
+def which(thefile):
+    path = os.environ.get("PATH", os.defpath).split(os.pathsep)
+    for dir in path:
+        name = os.path.join(dir, thefile)
+        if (os.path.exists(name) and os.access(name, os.F_OK | os.X_OK)
+                and not os.path.isdir(name)):
+            return name
+    return None
+
+
+def cmake_version(cmd):
+    for line in check_output([cmd, '--version']).split('\n'):
+        if 'version' in line:
+            return LooseVersion(line.strip().split(' ')[2])
+    raise Exception('no version found')
+
+
+def get_cmake_command():
+    cmake_command = 'cmake'
+    if IS_WINDOWS:
+        return cmake_command
+    cmake3 = which('cmake3')
+    if cmake3 is not None:
+        cmake = which('cmake')
+        if cmake is not None:
+            bare_version = cmake_version(cmake)
+            if bare_version < LooseVersion("3.5.0") and cmake_version(cmake3) > bare_version:
+                cmake_command = 'cmake3'
+    return cmake_command
+
+
+def cmake_defines(lst, **kwargs):
+    for key in sorted(kwargs.keys()):
+        value = kwargs[key]
+        if value is not None:
+            lst.append('-D{}={}'.format(key, value))
+
+
+# Ninja
+try:
+    import ninja
+    USE_NINJA = True
+except ImportError:
+    USE_NINJA = False
+
+base_dir = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+torch_lib_dir = base_dir + "/torch/lib"
+install_dir = base_dir + "/torch/lib/tmp_install"
+build_type = "Release"
+if DEBUG:
+    build_type = "Debug"
+elif REL_WITH_DEB_INFO:
+    build_type = "RelWithDebInfo"
+
+
+def mkdir_p(dir):
+    try:
+        os.makedirs(dir)
+    except OSError:
```

Comment:
One of the more annoying things about `makedirs`... I'd almost prefer the race-y version where just test if the dir exists first >.>